### PR TITLE
fix(schema): dedupe the generated schema artifacts with internal refs

### PR DESCRIPTION
**Pull request by hawkeyexl**
## Summary

The schema build fully dereferences every `$ref` and commits the results, duplicating shared components combinatorially — the element-criteria block appears ~200 times inside `step_v3.schema.json` alone, `schemas.json` sat at ~70 MB, the published `dist/index.cjs` inlined the same tens of megabytes, and one added component pushed the aggregate past GitHub's **100 MB file limit** during #691 (the push was rejected; that PR's schema component had to stay artificially terse).

This PR adds a **structural dedupe pass after dereferencing** (design record: ADR 01088):

- Repeated object subtrees (≥120 bytes, seen ≥2×) hoist into a reserved `x_dd_defs` container with internal `$ref`s; content-hash names keep output deterministic; definitions compress bottom-up so nesting dedupes maximally.
- **Self-verifying**: `dereferenceSchemas.cjs` expands what it just compressed and fails the build unless the result equals the dereferenced tree byte-for-byte.
- **Consumers never see the encoding**: `schemas/index.ts` expands at module load (AJV registration input unchanged), and the docs + type generators expand before walking.

## Results

| Artifact | Before | After |
|---|---|---|
| `schemas.json` | ~69.7 MB | **2.4 MB** |
| `resolvedTests_v3.schema.json` (largest) | ~22 MB | **322 KB** |
| `dist/index.cjs` (published CJS bundle) | ~70 MB inline | **2.4 MB** |

## Equivalence proof

- Expansion of the new committed artifacts **equals main's previously-committed artifacts exactly** for all 54 schemas (byte-level `JSON.stringify` comparison).
- Generated TypeScript types: regenerated, **zero diff**.
- Generated docs reference pages (`docs:build-schema-refs` + `docs:build-cli-ref`): regenerated, **zero diff**.
- src/common: 1,067 passing incl. the new `dedupe.test.js` round-trip/edge suite, **100% coverage ratchet holds**.
- Root: full suite **4,080 passing** with runtime `validate()` smoke-tested through both dist forms (ESM + CJS bundle).

## Notes

- The container key is deliberately not `definitions`/`$defs` (authored schemas own those — `config_v2` uses `definitions`); only a single-key `$ref` into `x_dd_defs` is treated as dedupe plumbing, and compression refuses inputs already using the namespace.
- The **published** raw `dist/schemas/*.json` files now carry the deduped encoding; package-export consumers are unaffected (they receive expanded objects). Anything reading the raw JSON files directly must resolve the internal refs (standard JSON pointers) or use the package export.
- No docs-page impact (reference pages byte-identical); ADR 01088 records the decision. Follow-up unlocked: #691's schema component can regain its schema-side `region` validation now that artifact size no longer constrains authoring.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Reduced stored schema artifact sizes via post-dereference structural deduplication.
  * Added automatic schema expansion so runtime consumers, CLI reference generation, and generated TypeScript types use fully inlined schemas.
  * Added verification to fail builds if compressed artifacts don’t expand back to the original schema.
* **Documentation**
  * Documented the deduplication/expansion format and compatibility expectations for existing schema namespaces.
* **Build Improvements**
  * Updated schema build/generation tooling and packaging to include the expansion/deduplication utilities in CommonJS and distribution outputs.
* **Tests**
  * Added coverage for compression/expansion correctness, determinism, and error/edge cases.

### CODE REVIEW

File: src/common/src/schemas/dereferenceSchemas.cjs [hunk 1]
```diff
@@ -129,10 +130,29 @@ async function dereferenceSchemas() {
...
+      fs.writeFileSync(outputFilePath, JSON.stringify(compressed, null, 2));
     } catch (err) {
       console.error(`Error processing ${file}:`, err);
+      // Per-file processing errors are historically tolerated (logged and
+      // skipped), but a failed dedupe INVARIANT means a wrong artifact
+      // would ship — that must fail the build.
+      if (String(err && err.message).includes("Dedupe round-trip mismatch")) {
+        throw err;
+      }
     }
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/common/src/schemas/dereferenceSchemas.cjs:165]: Fixed in 0b89bd7a — the compress/round-trip step is wrapped in its own try that tags every error as fatal, and the outer catch rethrows anything tagged, so namespace-collision guard errors (and any other dedupe failure) now fail the build instead of shipping a stale artifact. _🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/common/src/schemas/index.ts [hunk 2]
```diff
@@ -1,6 +1,13 @@
...
 export type Schema = (typeof schemasJson)[SchemaKey];
 
-export const schemas: typeof schemasJson = schemasJson;
+// The committed schemas.json is the DEDUPED on-disk encoding (repeated
+// subtrees hoisted into internal refs — see dedupe.cjs); consumers always
+// receive the expanded, fully-inlined form, exactly as the pre-dedupe
+// pipeline shipped it. Expansion happens once at module load.
+export const schemas: Record<SchemaKey, Schema> = Object.fromEntries(
+  Object.entries(schemasJson).map(([key, value]) => [key, expandSchema(value)])
+) as Record<SchemaKey, Schema>;
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/common/src/schemas/index.ts:17]: Fixed — `Schema` is now a generic JSON-object type (`{ [key: string]: unknown }`) so the public surface no longer describes the compressed x_dd_defs encoding; `SchemaKey` still derives from the committed JSON keys and the runtime export stays the expanded form. _🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/common/test/dedupe.test.js [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+      ).to.throw(/reserved/);
+    });
+
+    it("expand throws on an unresolvable container ref", function () {
+      expect(() =>
+        expandSchema({
+          properties: { x: { $ref: `#/${DEDUPE_CONTAINER}/missing` } },
+        })
+      ).to.throw(/unresolvable/i);
+    });
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/common/test/dedupe.test.js:139]: Added — a cyclic a→b→a container-ref test asserting the /cyclic/ throw (hand-corrupted-artifact scenario; compressSchema itself only produces trees). _🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
